### PR TITLE
docs/design: add proposal for invisible index

### PR DESCRIPTION
**Pull request by Deardrops**
### What problem does this PR solve? 
This PR adds a proposal for full  invisible index support in TiDB.

An issue has been created for the discussions on this proposal:

#9246

### CODE REVIEW

File: docs/design/2020-03-12-invisible-index.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+- Author(s):     [Deardrops](https://github.com/Deardrops), [wjhuang2016](https://github.com/wjhuang2016)
+- Last updated:  Mar. 12, 2020
+- Discussion at: https://github.com/pingcap/tidb/issues/9246
+
+## Abstract
+
+MySQL supports [Invisible indexes](https://dev.mysql.com/doc/refman/8.0/en/invisible-indexes.html); that is, indexes that are not used by the optimizer. 
+
+This is a useful feature when you want to drop an index in a safe way. Invisible indexes make it possible to test the effect of removing an index on query performance, without making a destructive change that must be undone should the index turn out to be required. Dropping and re-adding an index can be expensive for a large table, whereas making it invisible and visible are fast, in-place operations.
```
Comment (wjhuang2016) [inline docs/design/2020-03-12-invisible-index.md]: `should the index turn out to be required` ?

File: docs/design/2020-03-12-invisible-index.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+- Index hints need to report errors when used in invisible indexes.
+- The primary key cannot be set to an invisible index.
+
+## Background
+
+Index has a great influence on the performance of the database. Whether there is an index or whether the optimizer chooses the right index determines the performance of reading and writing the database to a great extent. In some cases, we would like to know the impact of deleting an index on the read-write performance of the database.  At present, our method is to tell the optimizer to ignore the index through Index Hint.  Although this method can achieve the goal, it is not practical to modify all SQL statements.
+
+## Proposal
+
+Adding an option (visible or not) to the index. If it is not visible, it's called **Invisible Index**. Invisible Index cannot be used by the optimizer (with the `use_invisible_indexes` switch on), but the index will be maintained during DML operations. For a query statement, invisible index has the same effect as ignoring the index through Index Hint.
```
Comment (wjhuang2016) [inline docs/design/2020-03-12-invisible-index.md]: but the index will be maintained during DML operations.      
s/will be/is

File: docs/design/2020-03-12-invisible-index.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+The invisible option of an index can be changed by using the following DDL statement:
```
Comment (wjhuang2016) [inline docs/design/2020-03-12-invisible-index.md]: s/invisible option/ option "invisible"

File: docs/design/2020-03-12-invisible-index.md (hunk 1 above)
Comment (Deardrops) [reply docs/design/2020-03-12-invisible-index.md]: This statement is copied from [mysql docs](https://dev.mysql.com/doc/refman/8.0/en/invisible-indexes.html), translating into Chinese:
```
不可见的索引可以测试移除索引对查询性能的影响，而不会做出破坏性的更改，如果（移除后）发现索引是必需的，则必须撤销这些更改。
```

**Comment by Deardrops** [comment]
@bb7133 This PR is ready for merge, PTAL

File: docs/design/2020-03-12-invisible-index.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
+# Proposal: Invisible index
+
+- Author(s):     [Deardrops](https://github.com/Deardrops), [wjhuang2016](https://github.com/wjhuang2016)
+- Last updated:  Mar. 12, 2020
+- Discussion at: https://github.com/pingcap/tidb/issues/9246
+
+## Abstract
+
+MySQL supports [Invisible indexes](https://dev.mysql.com/doc/refman/8.0/en/invisible-indexes.html); that is, indexes that are not used by the optimizer.
```
Comment (bb7133) [inline docs/design/2020-03-12-invisible-index.md]:
```suggestion
MySQL supports [invisible indexes](https://dev.mysql.com/doc/refman/8.0/en/invisible-indexes.html); that is, indexes that are not used by the optimizer. 
```

File: docs/design/2020-03-12-invisible-index.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+CREATE [...] INDEX index_name
+    [index_type]
+    ON tbl_name (key_part,...)
+    [index_option]
+
+index_option:
+    {VISIBLE | INVISIBLE}
+```
+
+Also, consider the following:
```
Comment (bb7133) [inline docs/design/2020-03-12-invisible-index.md]:
```suggestion
Also, the following behaviors need to be supported as well:
```

File: docs/design/2020-03-12-invisible-index.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+- Index hints need to report errors when used in invisible indexes.
+- The primary key cannot be set to an invisible index.
+
+## Background
+
+Index has a great influence on the performance of the database. Whether there is an index or whether the optimizer chooses the right index determines the performance of reading and writing the database to a great extent. In some cases, we would like to know the impact of deleting an index on the read-write performance of the database.  At present, our method is to tell the optimizer to ignore the index through Index Hint.  Although this method can achieve the goal, it is not practical to modify all SQL statements.
+
+## Proposal
+
+Adding an option (visible or not) to the index. If it is not visible, it's called **Invisible Index**. Invisible Index cannot be used by the optimizer (with the `use_invisible_indexes` switch on), but the index is maintained during DML operations. For a query statement, invisible index has the same effect as ignoring the index through Index Hint.
```
Comment (bb7133) [inline docs/design/2020-03-12-invisible-index.md]:
```suggestion
Adding an option (visible or not) to the index. If it is not visible, it's called **Invisible Index**. Invisible indexes cannot be used by the optimizer (with the `use_invisible_indexes` switch on), but the index is maintained during DML operations. For a query statement, invisible indexes have the same effect as ignoring the index through Index Hint.
```

File: docs/design/2020-03-12-invisible-index.md [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+The option `INVISIBLE` of an index can be changed by using the following DDL statement:
+
+```
+ALTER TABLE table_name ALTER INDEX index_name { INVISIBLE | VISIBLE };
+```
+
+Or set option `INVISIBLE` when creating an index:
```
Comment (bb7133) [inline docs/design/2020-03-12-invisible-index.md]:
```suggestion
Or by setting option `INVISIBLE` when creating an index:
```

File: docs/design/2020-03-12-invisible-index.md [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+Another solution for implement invisible indexes is: Indicate invisibility by DDL state `WriteOnly`. This solution has the following problems:
+
+- The logic of schema change needs to be changed, which is relatively complicated to implement.
+- Cannot distinguish between indexes that are currently in WriteOnly and invisible indexes.
+- Handling the switch is troublesome.
+
+## Compatibility and Migration Plan
+
+This the a new feature and it's absolutly compatible with old TiDB versions, also, it's not impact any data migration.
+The syntax and functions are basically compatible with MySQL. Expect one:
```
Comment (bb7133) [inline docs/design/2020-03-12-invisible-index.md]:
```suggestion
The syntax and functions are basically compatible with MySQL expect:
```

File: docs/design/2020-03-12-invisible-index.md [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+Another solution for implement invisible indexes is: Indicate invisibility by DDL state `WriteOnly`. This solution has the following problems:
+
+- The logic of schema change needs to be changed, which is relatively complicated to implement.
+- Cannot distinguish between indexes that are currently in WriteOnly and invisible indexes.
+- Handling the switch is troublesome.
+
+## Compatibility and Migration Plan
+
+This the a new feature and it's absolutly compatible with old TiDB versions, also, it's not impact any data migration.
```
Comment (bb7133) [inline docs/design/2020-03-12-invisible-index.md]:
```suggestion
This the a new feature and it's absolutly compatible with old TiDB versions, also, it does not impact any data migration.
```

File: docs/design/2020-03-12-invisible-index.md [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+The syntax and functions are basically compatible with MySQL. Expect one:
+
+	When use invisible index in `SQL Hint`, and set `use_invisible_indexes = false`, MySQL allow use the invisible index.
+	But in TiDB, It's **not allowed**, and will throw a `Unresolved name` error, Because the behavior is more reasonable.
```
Comment (bb7133) [inline docs/design/2020-03-12-invisible-index.md]:
```suggestion
	But in TiDB, It's **not allowed** and an `Unresolved name` error will be thrown.
```

File: docs/design/2020-03-12-invisible-index.md [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+- Add a new column `VISIBLE` in `SHOW INDEX FROM table` statement
+- Show invisiable column infomations in `SHOW CREATE TABLE` statement
+- Add `use_invisible_indexes` in system variable `@@optimizer_switch`
+- Add new error message `ERROR 3522 (HY000): A primary key index cannot be invisible`
+- Ignore invisible index in optimizer and add unit tests
+
+## Testing Plan
+
+- Unit test
+- Port All mysql test related to invisible index
```
Comment (bb7133) [inline docs/design/2020-03-12-invisible-index.md]:
```suggestion
- Learn from MySQL test cases related to invisible indexes
```

**Pull request merged**